### ISSUE TITLE
The --static-dir flag is not respected

### ISSUE THREAD

**Issue by alerque** [knadh/listmonk#340]
The `--static-dir` CLI flag seems to be completely ignored. I can pass it a path to the static resources and it throws an error about not being able to find them. If instead I `cd` into the static resources dir I previously mentioned as an arg and skip the arg it finds the resource files.

```console
$ cd /tmp

$ sudo -u listmonk listmonk --config /etc/listmonk/config.toml --static-dir /usr/share/listmonk
2021/05/05 16:10:55 init.go:123: using local filesystem for static assets
2021/05/05 16:10:55 init.go:143: failed to initialize local file for assets: stat config.toml.sample: permission denied

$ cd /usr/share/listmonk

$ sudo -u listmonk listmonk --config /etc/listmonk/config.toml
2021/05/05 16:13:05 init.go:123: using local filesystem for static assets
```

The latter invocation works, but the former should have.

**Comment by knadh** [comment]
Definitely something to do with permissions on your system. This feature works. Here's it running from the binary downloaded from the releases page. Custom config path and custom static dir path.

```
$ ./listmonk --config=../config.toml --static-dir=/tmp/listmonk/static
2021/05/06 12:42:03 main.go:84: v0.9.0-beta (e90fb1d 2021-02-13T12:41:40Z)
2021/05/06 12:42:03 init.go:98: reading config: ../config.toml
2021/05/06 12:42:03 init.go:175: connecting to db: localhost:9432/listmonk
2021/05/06 12:42:03 init.go:149: loading static files from: /tmp/listmonk/static
2021/05/06 12:42:03 init.go:420: media upload provider: filesystem
2021/05/06 12:42:03 init.go:344: loaded email (SMTP) messenger: username@smtp.yoursite.com
⇨ http server started on [::]:9000

```

**Comment by alerque** [comment]
@knadh I'm using a pretty bone-stock Linux setup, no special ACLs etc.

By the way this whole procedure is for setting up packaging on Arch Linux. I'm testing on several systems including stripped down minimal systems just making sure this package works.

Can you show a few more details from your working arrangement? Maybe we can spot the difference. First of all I'm suspicious of you running `./listmonk`. That suggests you are perhaps running inside the source directory where you compiled it, and almost certainly running as the same user that owns the binary. Your `/tmp` dir for statics is probably owned as the same as your UID & the binary owner too. Those should not be prerequisites.

Lets see the output of these:

```console
$ stat ./listmonk /tmp/listmonk/static $(pwd)
$ id
```

For context here are relevant lines from my system(s). The difference I expect to see that is not reflected in your testing is that the owner of the binary and static files is the system, but running as a non-privileged user (with `rx` permission on the binary and `r` on the statics (+`x` for dir).

```console
$ stat $(which listmonk) /usr/share/listmonk $(pwd)
  File: /usr/bin/listmonk
  Size: 15922472        Blocks: 31104      IO Block: 4096   regular file
Device: 1ah/26d Inode: 440712      Links: 1
Access: (0755/-rwxr-xr-x)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2021-05-05 20:44:33.343520940 +0200
Modify: 2021-05-05 20:30:41.000000000 +0200
Change: 2021-05-05 20:43:25.863573805 +0200
 Birth: 2021-05-05 20:43:25.833573829 +0200
  File: /usr/share/listmonk
  Size: 114             Blocks: 0          IO Block: 4096   directory
Device: 1ah/26d Inode: 440715      Links: 1
Access: (0755/drwxr-xr-x)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2021-05-05 20:46:33.263436360 +0200
Modify: 2021-05-05 20:43:25.873573797 +0200
Change: 2021-05-05 20:43:25.873573797 +0200
 Birth: 2021-05-05 20:43:25.863573805 +0200
  File: /
  Size: 218             Blocks: 0          IO Block: 4096   directory
Device: 1ah/26d Inode: 256         Links: 1
Access: (0755/drwxr-xr-x)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2021-05-05 12:19:01.766151913 +0200
Modify: 2021-04-22 00:20:26.168645841 +0200
Change: 2021-04-22 00:20:26.168645841 +0200
 Birth: 2021-04-21 23:19:50.799860474 +0200
 $ id
uid=966(listmonk) gid=966(listmonk) groups=966(listmonk)
```

Can you confirm that it still works for you if the binary and statics dir are owned by root but run by your user?

**Comment by alerque** [comment]
Another clue, although I don't really understand how to interpret it, is that I'm seeing this error even when running from inside the statics folder (so a fully working system that I ran a test campain on). Specifying the statics folder should probably keep the binary from even *trying* to access embedded files but it evidently does not.

```
2021/05/06 10:17:54 init.go:123: unable to initialize embedded filesystem: no ID found in the file
```

**Comment by alerque** [comment]
New idea: this bug is probably related to *not* using [`stuffbin`](https://github.com/knadh/stuffbin) to embed assets. These builds are expected to use a static directory to serve assets. The binary is being used after being built by `go` without the post-processing that the upstream `make dist` or `make pack-bin` steps take.

Other than this snafu it seems to work fine that way. The service runs, the admin site is served, mail campaigns run, etc.

Apparently when it fails to find the embed resources it also fails to process the CLI flag for where to find them and *only* looks at the current working directory.

**Comment by knadh** [comment]
> First of all I'm suspicious of you running ./listmonk. That suggests you are perhaps running inside the source directory where you compiled it

That would be a futile test! As I said, it's a fresh binary downloaded from the releases page (and as I already said, this release has been running in production in many places.)


Binary in a directory.

```shell
kailash@nova (/tmp/listmonk_0.9.0-beta_linux_amd64)
$ ls -lh
total 12M
-rwxr-xr-x. 1 kailash kailash 12M Feb 13 18:11 listmonk
```

Config and static dir in another.

```shell
kailash@nova (/tmp/listmonk)
$ ls -lh
total 4.0K
-rw-r--r--. 1 kailash kailash 288 May  6 15:45 config.toml
drwxr-xr-x. 4 kailash kailash  80 May  6 15:43 static
```

Putting it all together.

```shell
kailash@nova (/home/kailash)
$ /tmp/listmonk_0.9.0-beta_linux_amd64/listmonk --config=/tmp/listmonk/config.toml --static-dir=/home/kailash/code/go/my/knadh/listmonk/static

2021/05/06 15:45:45 main.go:84: v0.9.0-beta (e90fb1d 2021-02-13T12:41:40Z)
2021/05/06 15:45:45 init.go:98: reading config: /tmp/listmonk/config.toml
2021/05/06 15:45:45 init.go:175: connecting to db: localhost:9432/listmonk
2021/05/06 15:45:45 init.go:149: loading static files from: /home/kailash/code/go/my/knadh/listmonk/static
2021/05/06 15:45:45 init.go:420: media upload provider: filesystem
2021/05/06 15:45:45 init.go:344: loaded email (SMTP) messenger: username@smtp.yoursite.com
⇨ http server started on [::]:9000
```


> Your /tmp dir for statics is probably owned as the same as your UID & the binary owner too.

Yes indeed. As the above example shows, the binary indeed has static files embedded in it and works as expected. Now if uid/permissions cause the binary to not be able to read itself (as I've been suggesting), that's a different problem altogether, unrelated to the release and the binary.


> Specifying the statics folder should probably keep the binary from even trying to access embedded files but it evidently does not.

This is by design and allows you to override individual static files without having to copy and maintain the entire tree. Files picked up from the external static dir ar "merged" into whatever is in memory.

**Comment by knadh** [comment]
> New idea: this bug is probably related to _not_ using [`stuffbin`](https://github.com/knadh/stuffbin) to embed assets. These builds are expected to use a static directory to serve assets. The binary is being used after being built by `go` without the post-processing that the upstream `make dist` or `make pack-bin` steps take.

Ah! All this while, I assumed you were referring to the pre-built binary on the releases page! If you're building locally, you've to indeed use stuffbin to embed the static files post go build. `make dist` does this.

**Comment by alerque** [comment]
For the record, this is a different issue from #338. I should have clarified, but:

* the issue with files in the static bin compiled upstream not being accessible when run under a different user is still an issue (see #338)
* this issue about the CLI flag being ignored when the embedded file set is not included and *everything* should be coming from a specified static directory is a different issue.

Respectfully even if my use case in distro packaging is contrary to your expectations, the fact that it 95% works and only fails in an unexpected way at run time is a bug.

Here is how I think it should be fixed:

* The main binary should not throw an error if there are no embedded files, maybe a warning maybe not.
* The CLI flag handling for `--static-dir` should work whether or not embedded files are found at all.
* If no embeded files are found, the `--static-dir` argument should actually be mandatory and a check that the directory exists and is readable should replace the check for internal files.

As a bonus measure:

* The config file should recognize a `static_dir` option and respect it in place of the CLI flag. This would make systems with default Systemd services easier to customize.

**Comment by knadh** [comment]
> the issue with files in the static bin compiled upstream not being accessible when run under a different user is still an issue (see #338)

But this has nothing to do with listmonk or the way it embeds static files though. It's merely doing a `os.Open()`, a simple file open, [on itself](https://github.com/knadh/stuffbin/blob/e80f23deca3fbc201ae7fb5f59a9e4ea6f17878e/stuff.go#L75). That in a certain environment under certain kind of permissions, `os.Open()` fails to open a path is really an external problem.

> this issue about the CLI flag being ignored when the embedded file set is not included and everything should be coming from a specified static directory is a different issue.

This is also not an issue and again, seems to do with permissions. Below is a binary with no embedded files running, picking up local static files as a fallback, then respecting `--static-dir` and loading it from `/tmp/listmonk/static`

```shell
kailash@nova (/listmonk) ● master                                                                                            
$ ./listmonk --static-dir=/tmp/listmonk/static/                                                                                                            
2021/05/06 17:33:06 main.go:84: v0.9.0-beta (#68b80d0 2021-05-06T12:03:04+0000)
2021/05/06 17:33:06 init.go:99: reading config: config.toml
2021/05/06 17:33:06 init.go:189: connecting to db: localhost:9432/listmonk
2021/05/06 17:33:06 init.go:123: unable to initialize embedded filesystem: no ID found in the file
2021/05/06 17:33:06 init.go:124: using local filesystem for static assets
2021/05/06 17:33:06 init.go:150: loading static files from: /tmp/listmonk/static/
2021/05/06 17:33:06 init.go:438: media upload provider: filesystem
2021/05/06 17:33:06 init.go:362: loaded email (SMTP) messenger: username@smtp.yoursite.com
⇨ http server started on [::]:9000
```

> The main binary should not throw an error if there are no embedded files, maybe a warning maybe not.

It doesn't. It merely warns. See above.

> The CLI flag handling for --static-dir should work whether or not embedded files are found at all.

It does. See above.

> If no embeded files are found, the --static-dir argument should actually be mandatory and a check that the directory exists and is readable should replace the check for internal files.

It fails right correctly, but yes, I agree, the error message can specifically mandate `--static-dir`.

> The config file should recognize a static_dir option and respect it in place of the CLI flag. This would make systems with default Systemd services easier to customize.

It does (yep, this needs to be documented). Putting `static-dir` in the config.toml outside of the root `[app]` works.

**Comment by knadh** [comment]
> The CLI flag handling for --static-dir should work whether or not embedded files are found at all.

Correction: I see what you mean. It works even if no embedded files are found, but not if the static dir is not in the path. You are right, this needs to be fixed.

**Comment by alerque** [comment]
> > The config file should recognize a static_dir option and respect it in place of the CLI flag. This would make systems with default Systemd services easier to customize.
> 
> It does (yep, this needs to be documented). Putting `static-dir` in the config.toml outside of the root `[app]` works.

Okay I had tried inside [app], it didn't occur to me to try it outside. Documentation would fix that bit.

**Comment by knadh** [comment]
On a different note, Go 1.16 has built in support for [embedding files](https://golang.org/doc/go1.16#library-embed), negating the need for an external tool like stuffbin. I haven't had a chance to look at this, but once this is implemented, static file embedded will become an inherent part of `go build` and solve this issue for good. I'll explore this.

**Comment by alerque** [comment]
I appreciate the attempt to fix this, but in my testing of the 1.0.0 release it is not actually fixed.

Mind you the console log message are different, but it doesn't actually work.

1. From *inside* the /usr/share/listmonk directory
   - launching works with no argument as before
   - launching with `--static-dir /usr/share/listmonk/static` echos a message to the console that suggests it is using that path (which is different than with no argument
   - launching with `--static-dir /usr/share/listmonk` fails saying it can't find "email-templates" suggesting that this argument really does want the "static" folder, not the parent directory where it currently needs to be launched.
   - launching with `--stataic-dir /foo` fails with an error, suggesting that it's actually using that to load email tempaletes
1. From any other PWD, e.g. `cd /tmp`:
   - launching with no argument fails (as before)
   - launching with `--static-dir /usr/share/listmonk/static` fails with an error that "config.toml.sample" does not exist. First of all, given that I'm also passing `--config /etc/listmonk/config.toml` who cares about the sample file! Why should it even be looking? Second this file isn't in the "static" directory, it's in one directory up from there, suggesting the argument doesn't know what it wants to be.
   - launching with `--static-dir /usr/share/listmonk` fails with an error that "config.toml.sample" does not exist. Again it still doesn't find this file, which also suggests it wont' be able to find the i18n or sql files there either.
1. No iteration of setting `static_dir` in the config file seems to have any affect as was suggested in earlier comments. The log messages don't reflect the same results and the `--static-dir` CLI arg, they reflect the same results as not having an argument at all.

This needs more work:

1. Check the CLI flags first, then the config file.
1. Stop looking for sample config files if it's not used for anything
1. Add another setting if necessary for where the run time files like i18n that are that are not "statics" in the sense of user-editable templates. Perhaps statics should be renamed to "templates" and a new setting for other runtime assets that should not be edited but always come with the package should be added.
1. Use static assets from the specific setting if present, then fall back to the runtime directory (can be a distro installed bath) or lastly the embedded system. It should be possible for users to customize this without changing anything in their distro's installed path while still using a file system path for distro packages (no embedded files).
